### PR TITLE
Grpc/support groupby with gen and rerank

### PR DESCRIPTION
**Pull request by tsmith023**
### Review checklist

- [ ] Documentation has been updated, if necessary. Link to changed documentation:
- [ ] Chaos pipeline run or not necessary. Link to pipeline:
- [x] All new code is covered by tests where it is reasonable.
- [ ] Performance tests have been run or not necessary.

### CODE REVIEW

File: adapters/handlers/grpc/v1/prepare_reply.go [hunk 1]
```diff
@@ -311,6 +317,56 @@ func extractGroup(raw any, searchParams dto.GetParams, scheme schema.Schema, use
...
+	}
+
+	groupedGenerativeResults := ""
+	if generativeSearchEnabled {
+		generate, ok := addAsMap["generate"]
+		if !ok {
+			return nil, "", errors.New("No results for generative search despite a search request. Is a the generative module enabled?")
+		}
+
+		if ok {
```
Comment (dirkkul) [inline adapters/handlers/grpc/v1/prepare_reply.go:334]: we don't need the `if ok` - you just did `if !ok` above

File: adapters/handlers/grpc/v1/prepare_reply.go [hunk 2]
```diff
@@ -34,7 +34,8 @@ import (
 func searchResultsToProto(res []interface{}, start time.Time, searchParams dto.GetParams, scheme schema.Schema, usesPropertiesMessage bool) (*pb.SearchReply, error) {
 	tookSeconds := float64(time.Since(start)) / float64(time.Second)
 	out := &pb.SearchReply{
-		Took: float32(tookSeconds),
+		Took:                    float32(tookSeconds),
+		GenerativeGroupedResult: new(string), // pointer to empty string
```
Comment (dirkkul) [inline adapters/handlers/grpc/v1/prepare_reply.go:38]: why is this needed? Can't we just leave it `nil`?

File: adapters/handlers/grpc/v1/prepare_reply.go [hunk 3]
```diff
@@ -290,6 +293,9 @@ func extractAdditionalProps(asMap map[string]any, additionalPropsParams addition
 }
 
 func extractGroup(raw any, searchParams dto.GetParams, scheme schema.Schema, usesMarshalling bool) (*pb.GroupByResult, string, error) {
+	_, generativeSearchEnabled := searchParams.AdditionalProperties.ModuleParams["generate"]
+	_, rerankEnabled := searchParams.AdditionalProperties.ModuleParams["rerank"]
```
Comment (dirkkul) [inline adapters/handlers/grpc/v1/prepare_reply.go:297]: move this a level up so we do not need to do this for every object?

File: adapters/handlers/grpc/v1/prepare_reply.go [hunk 4]
```diff
@@ -311,6 +317,56 @@ func extractGroup(raw any, searchParams dto.GetParams, scheme schema.Schema, use
...
+			}
+
+			if generateFmt.SingleResult != nil && *generateFmt.SingleResult != "" {
+				ret.Generative = &pb.GenerativeReply{Result: *generateFmt.SingleResult}
+			}
+
+			// grouped results are only added to the first object for GQL reasons
+			// however, reranking can result in a different order, so we need to check every object
+			// recording the result if it's present assuming that it is at least somewhere and will be caught
+			if generateFmt.GroupedResult != nil && *generateFmt.GroupedResult != "" {
```
Comment (dirkkul) [inline adapters/handlers/grpc/v1/prepare_reply.go:350]: should we add a check to the level above to make sure that it was really caught?

File: adapters/handlers/grpc/v1/prepare_reply.go [hunk 5]
```diff
@@ -327,12 +383,12 @@ func extractGroup(raw any, searchParams dto.GetParams, scheme schema.Schema, use
 		returnObjectsUntyped[i] = group.Hits[i]
 	}
 
-	objects, groupedGenerativeResults, err := extractObjectsToResults(returnObjectsUntyped, searchParams, scheme, true, usesMarshalling)
+	objects, _, err := extractObjectsToResults(returnObjectsUntyped, searchParams, scheme, true, usesMarshalling)
```
Comment (dirkkul) [inline adapters/handlers/grpc/v1/prepare_reply.go:386]: why this change?

**Review by dirkkul — COMMENTED** [review]
any tests?

File: adapters/handlers/grpc/v1/prepare_reply.go (hunk 2 above)
Comment (tsmith023) [reply adapters/handlers/grpc/v1/prepare_reply.go:38]: Not having it broke a unit the `group by` unit test in `prepare_reply_test.go` on L976 due to a nil pointer deference

In addition, gRPC field is `optional string` and the previous behaviour was for it to always be `""` so changing this would be a BC change I think

File: adapters/handlers/grpc/v1/prepare_reply.go (hunk 3 above)
Comment (tsmith023) [reply adapters/handlers/grpc/v1/prepare_reply.go:297]: Ideally, this would also be the case for `extractObjectsToResults` but that would be a large refactor beyond this PR's scope

File: adapters/handlers/grpc/v1/prepare_reply.go (hunk 5 above)
Comment (tsmith023) [reply adapters/handlers/grpc/v1/prepare_reply.go:386]: The generative result in a group by query doesn't come from the `hits` field of the map but instead from the `generate` field on the same level as the `hits` field. Since `extractObjectsToResults` here is acting on `returnObjectsUntyped`, which is a list of `group.Hits` then it cannot have generated results in it

**Comment by tsmith023** [comment]
>any tests?

Will add some unit tests to `prepare_reply_test.go`

**Review by dirkkul — APPROVED** [review]
